lib/config/presets.js

```diff
@@ -0,0 +1,192 @@
+const configParser = require('./index');
+const massage = require('./massage');
+const npm = require('../api/npm');
+
+module.exports = {
+  resolveConfigPresets,
+  replaceArgs,
+  parsePreset,
+  getPreset,
+};
+
+async function resolveConfigPresets(
+  inputConfig,
+  logger = inputConfig.logger,
+  existingPresets = []
+) {
+  logger.trace(
+    { config: inputConfig, existingPresets },
+    'resolveConfigPresets'
+  );
+  let config = {};
+  // First, merge all the preset configs from left to right
+  if (inputConfig.extends) {
+    logger.debug('Found presets');
+    for (const preset of inputConfig.extends) {
+      // istanbul ignore if
+      if (existingPresets.indexOf(preset) !== -1) {
+        logger.warn(`Already seen preset ${preset} in ${existingPresets}`);
+      } else {
+        logger.debug(`Resolving preset "${preset}"`);
+        const presetConfig = await resolveConfigPresets(
+          await getPreset(preset, logger),
+          logger,
+          existingPresets.concat([preset])
+        );
+        config = configParser.mergeChildConfig(config, presetConfig);
+      }
+    }
+  }
+  logger.trace({ config }, `Post-preset resolve config`);
+  // Now assign "regular" config on top
+  config = configParser.mergeChildConfig(config, inputConfig);
+  delete config.extends;
+  logger.trace({ config }, `Post-merge resolve config`);
+  for (const key of Object.keys(config)) {
+    const val = config[key];
+    if (isObject(val) && key !== 'logger') {
+      // Resolve nested objects
+      logger.trace(`Resolving object "${key}"`);
+      config[key] = await resolveConfigPresets(val, logger, existingPresets);
+    } else if (Array.isArray(val)) {
+      // Resolve nested objects inside arrays
+      config[key] = [];
+      for (const element of val) {
+        if (isObject(element)) {
+          config[key].push(
+            await resolveConfigPresets(element, logger, existingPresets)
+          );
+        } else {
+          config[key].push(element);
+        }
+      }
+    }
+  }
+  logger.debug({ config: inputConfig }, 'Input config');
+  logger.debug({ config }, 'Resolved config');
+  return config;
+}
+
+function replaceArgs(obj, argMapping) {
+  if (typeof obj === 'string') {
+    let returnStr = obj;
+    for (const arg of Object.keys(argMapping)) {
+      returnStr = returnStr.replace(`{{${arg}}}`, argMapping[arg]);
```

Comment:
I'm not sure if we need to use `{{argX}}` twice or more (maybe in description?), but it'd be better to replace all the matched target instead of just the first one.